sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -80,10 +80,6 @@ EventImplPtr Scheduler::addCG(std::unique_ptr<detail::CG> CommandGroup,
     default:
       NewCmd = MGraphBuilder.addCG(std::move(CommandGroup), std::move(Queue));
     }
-  }
-
-  {
-    std::shared_lock<std::shared_timed_mutex> Lock(MGraphLock);
```

Comment:
There are also some other places which enqueue commands for execution. Currently they are using read-lock. I believe we can face the same issue there as well. Also, your intent here is to serialize enqueue operations. Well, the target of your intent is not achieved yet as enqueue operations may be executed in parallel in that places.

Reply:
Yes, I checked all other places which enqueue commands for execution and faced problem the Lit tests hangs, you can have a look at previous commits and build bot results.
So, the Current version passed all Lit tests and fixed the problem with the regression test.

Reply:
I believe there must be some solution here.
As far as I understood the changes have caused a deadlock. What's the purpose of write-lock? Is it employed here to prevent simultaneous changes to structure of graph or changes to some internal data of commands in graph? We can get by the latter with employing a per-command mutex or atomic property with proper combination of CAS/read/exchange operations.

Reply:
Yes, I investigated the root cause of the problem and found the race when setting kernel parameters in parallel. That occurs when submitting to a single queue by multiple threads. So I added per-kernel mutex.